src/goto-instrument/contracts/contracts.cpp

```diff
@@ -252,13 +252,30 @@ void code_contractst::check_apply_loop_contracts(
     loop_head,
     add_pragma_disable_assigns_check(generated_code));
 
-  // havoc the variables that may be modified
-  assignst assigns;
+  assignst to_havoc;
+
   if(assigns_clause.is_nil())
   {
+    // No assigns clause was specified for this loop.
+    // Infer memory locations assigned by the loop from the loop instructions
+    // and the inferred alisaing relation.
     try
     {
-      get_assigns(local_may_alias, loop, assigns);
+      get_assigns(local_may_alias, loop, to_havoc);
+      log.warning() << "No loop assigns clause provided. Inferred targets {";
+      // Add inferred targets to the loop assigns clause.
+      bool ran_once = false;
+      for(auto &target : to_havoc)
```

Comment:
Probably `const auto &` to avoid unnecessary copies in case some of loop body could (also) be invoked in the non-const context and therefore trigger copy-on-write.